// unplugin-compression/dist/shared/unplugin-compression.d8d20b98.mjs
import { createUnplugin } from 'unplugin';
import * as compressing from 'compressing';
import path, { resolve } from 'path';
import fs from 'fs/promises';
import c from 'picocolors';

class Log {
  static log(msg) {
    console.log(
      `${c.inverse(c.bold(c.magenta("UNPLUGIN COMPRESSING")))} ${msg}`
    );
  }
  static error(msg) {
    console.log(
      `${c.inverse(c.bold(c.red("UNPLUGIN COMPRESSING")))} ${c.red(msg)}`
    );
  }
  static success(msg) {
    console.log(
      `${c.inverse(c.bold(c.green("UNPLUGIN COMPRESSING")))} ${c.green(msg)}`
    );
  }
}

function normalizerSource(source, globalOptions) {
  if (typeof source === "string") {
    source = {
      source
    };
  }
  if (!Array.isArray(source)) {
    source = [source];
  }
  source = source.map((s) => {
    const defaultFormatter = "{{name}}.{{ext}}";
    return {
      source: s.source ?? "dist",
      outDir: s.outDir ?? globalOptions.outDir ?? "./",
      adapter: s.adapter ?? globalOptions.adapter ?? "zip",
      formatter: s.formatter ?? globalOptions.formatter ?? defaultFormatter,
      compressingOptions: s.compressingOptions ?? globalOptions.compressingOptions ?? void 0
    };
  });
  return source;
}
function getCompressName(source) {
  const { formatter, ...s } = source;
  const extra = {
    name: path.basename(s.source),
    ext: s.adapter
  };
  if (typeof formatter === "string") {
    const template = { ...s, ...extra };
    return formatter.replace(/\{\{(\w+)\}\}/g, ($_, $1) => {
      const value = template[$1];
      if (value !== void 0) {
        return value;
      }
      return $_;
    });
  }
  return formatter(s);
}
async function compress({
  source: rawSource = "dist",
  ...options
} = {}) {
  const source = normalizerSource(rawSource, options);
  await Promise.all(
    source.map(async (s) => {
      const compressAdapter = compressing[s.adapter];
      const resolvedSource = resolve(process.cwd(), s.source);
      const fileName = getCompressName(s);
      const resolvedOutput = path.join(
        resolve(process.cwd(), s.outDir),
        fileName
      );
      await compressAdapter.compressDir(resolvedSource, resolvedOutput, s.compressingOptions);
      const stat = await fs.stat(resolvedOutput);
      Log.success(`${fileName}	${stat.size} bytes`);
    })
  );
  return true;
}

const unplugin = createUnplugin((options) => ({
  name: "unplugin-compress",
  enforce: "post",
  buildEnd() {
    let isCompress = false;
    process.on("beforeExit", async () => {
      if (isCompress)
        return;
      isCompress = true;
      Log.log("start compressing");
      let compressed = false;
      try {
        compressed = await compress(options);
      } catch (e) {
        Log.error(String(e));
      }
      if (compressed) {
        Log.success("compressed success");
      }
    });
  }
}));

export { unplugin as u };
